lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -268,137 +268,224 @@ private static Domain getDomain(
+                        long minValue = min instanceof Binary ? getShortDecimalValue(((Binary) min).getBytes()) : asLong(min);
+                        long maxValue = max instanceof Binary ? getShortDecimalValue(((Binary) max).getBytes()) : asLong(max);
 
-                    if (isStatisticsOverflow(type, minValue, maxValue)) {
-                        return Domain.create(ValueSet.all(type), hasNullValue);
+                        if (isStatisticsOverflow(type, minValue, maxValue)) {
+                            return Domain.create(ValueSet.all(type), hasNullValue);
+                        }
+
+                        ranges.add(Range.range(type, minValue, true, maxValue, true));
                     }
+                    return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
+                }
 
-                    ranges.add(Range.range(type, minValue, true, maxValue, true));
+                DiscreteValuesList discreteValuesList = (DiscreteValuesList) valuesList;
+                List<Long> values = new ArrayList<>(valuesList.size());
+                for (Object value : discreteValuesList.values()) {
+                    long valueAsLong = value instanceof Binary ? getShortDecimalValue(((Binary) value).getBytes()) : asLong(value);
+                    if (isStatisticsOverflow(type, valueAsLong, valueAsLong)) {
+                        return Domain.create(ValueSet.all(type), hasNullValue);
+                    }
+                    values.add(valueAsLong);
                 }
+                return Domain.create(ValueSet.copyOf(type, values), hasNullValue);
             }
-            else {
-                for (int i = 0; i < minimums.size(); i++) {
-                    Int128 min = Int128.fromBigEndian(((Binary) minimums.get(i)).getBytes());
-                    Int128 max = Int128.fromBigEndian(((Binary) maximums.get(i)).getBytes());
+
+            if (valuesList instanceof StatisticsRangeList statisticsRanges) {
+                List<Range> ranges = new ArrayList<>(valuesList.size());
+                for (StatisticsRange statisticsRange : statisticsRanges.rangesList()) {
+                    Int128 min = Int128.fromBigEndian(((Binary) statisticsRange.min()).getBytes());
+                    Int128 max = Int128.fromBigEndian(((Binary) statisticsRange.max()).getBytes());
 
                     ranges.add(Range.range(type, min, true, max, true));
                 }
+                return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
             }
 
-            return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
+            DiscreteValuesList discreteValuesList = (DiscreteValuesList) valuesList;
+            List<Int128> values = new ArrayList<>(valuesList.size());
+            for (Object value : discreteValuesList.values()) {
+                values.add(Int128.fromBigEndian(((Binary) value).getBytes()));
+            }
+            return Domain.create(ValueSet.copyOf(type, values), hasNullValue);
```

Comment:
I think you could have a bit less code if you checked `valuesList` first and then `decimal.isShort()`
the
```
List<Object> values = new ArrayList<>(valuesList.size());
...
return Domain.create(ValueSet.copyOf(type, values), hasNullValue);
```

would be shared.

Reply:
Trying to share that makes it slightly harder to read because I have to
```
if (decimal.isShort()) {
  .....
}
else {
  .....
}
```
instead of
```
if (decimal.isShort()) {
  ......
}
....
```